cupyx/jit/_compile.py

```diff
@@ -195,6 +227,21 @@ def _transpile_function(
         env (Environment): More details of analysis result of the function,
             which includes preambles, estimated return type and more.
     """
+    try:
+        return _transpile_function_internal(
+            func, attributes, mode, consts, in_types, ret_type)
+    except JitCompileError as e:
+        exc = e
+        if _is_debug_mode:
+            exc.reraise(source)
+
+    # Raises the error out of `except` block to clean stack trace.
+    exc.reraise(source)
```

Comment:
I am confused by the purpose: Doesn't this mean we always raise the JitCompileError if it happens? Would the stack trace depends on whether we raise it inside the `except` block or not?

Reply:
Updated the PR description 😃

Reply:
If reraised the error inside the `except` block, the stack traces of internal functions `_transpile_{stmt/expr}` are also printed, but it is annoying and not benefical for the users. So I implemented to reraise the error outside of `except` block.

Reply:
Wow I didn't know this trick, cool!

Before:
```python
Traceback (most recent call last):
  File "test_jit_error.py", line 8, in <module>
    f[1, 1]()
  File "/home/leofang/dev/cupy_cuda112/cupyx/jit/_interface.py", line 82, in <lambda>
    return lambda *args, **kwargs: self(grid, block, args, **kwargs)
  File "/home/leofang/dev/cupy_cuda112/cupyx/jit/_interface.py", line 66, in __call__
    _types.Void(),
  File "/home/leofang/dev/cupy_cuda112/cupyx/jit/_compile.py", line 51, in transpile
    tree.body[0], attributes, mode, consts, in_types, ret_type)
  File "/home/leofang/dev/cupy_cuda112/cupyx/jit/_compile.py", line 231, in _transpile_function
    body = _transpile_stmts(func.body, True, env)
  File "/home/leofang/dev/cupy_cuda112/cupyx/jit/_compile.py", line 318, in _transpile_stmts
    codeblocks.extend(_transpile_stmt(stmt, is_toplevel, env))
  File "/home/leofang/dev/cupy_cuda112/cupyx/jit/_compile.py", line 334, in _transpile_stmt
    value = _transpile_expr(stmt.value, env)
  File "/home/leofang/dev/cupy_cuda112/cupyx/jit/_compile.py", line 464, in _transpile_expr
    res = _transpile_expr_internal(expr, env)
  File "/home/leofang/dev/cupy_cuda112/cupyx/jit/_compile.py", line 642, in _transpile_expr_internal
    f'Unbound name: {expr.id} in line {expr.lineno}')
NameError: Unbound name: abc in line 3
```
After:
```python
Traceback (most recent call last):
  File "test_jit_error.py", line 8, in <module>
    f[1, 1]()
  File "/home/leofang/dev/cupy_cuda112/cupyx/jit/_interface.py", line 82, in <lambda>
    return lambda *args, **kwargs: self(grid, block, args, **kwargs)
  File "/home/leofang/dev/cupy_cuda112/cupyx/jit/_interface.py", line 66, in __call__
    _types.Void(),
  File "/home/leofang/dev/cupy_cuda112/cupyx/jit/_compile.py", line 82, in transpile
    source=source
  File "/home/leofang/dev/cupy_cuda112/cupyx/jit/_compile.py", line 239, in _transpile_function
    exc.reraise(source)
  File "/home/leofang/dev/cupy_cuda112/cupyx/jit/_compile.py", line 37, in reraise
    raise self.error_type(self.mes + '\n\n' + pycode)
NameError: Unbound name: abc

  @cupyx.jit.rawkernel()
  def f():
>     return abc

```